### ISSUE TITLE
Non-optimal walking route outcome

### ISSUE THREAD

**Issue by PinkDuck** [osmandapp/Osmand#3575]
Can be tested with OsmAnd Map Creator, using routing=pedestrian and log file.

With added transit point to shorten at 52.631462, 1.2312781:
```
INFO: <test regions="" description="" best_percent="" vehicle="pedestrian" 
    start_lat="52.62801746570941" start_lon="1.2347488379478193" target_lat="52.64364574368104" target_lon="1.21101128816602" loadedTiles = "121" visitedSegments = "954" complete_distance = "2996.7266" complete_time = "2227.643" routing_time = "2152.0234"  >
```

Default pedestrian route without:
```
INFO: <test regions="" description="" best_percent="" vehicle="pedestrian" 
    start_lat="52.62801746570941" start_lon="1.2347488379478193" target_lat="52.64364574368104" target_lon="1.21101128816602" loadedTiles = "85" visitedSegments = "1543" complete_distance = "3007.7664" complete_time = "2235.5918" routing_time = "2159.6848"  >
```

Which is 11m longer, 8 seconds complete time longer, 7.7 seconds routing time longer.

OBF files used uploaded to [OneDrive](https://1drv.ms/u/s!Ah1Oa_WAIEBgxXeGjSSoRxqp6VAk) (29.8 MB zip archive)

I'm not sure why this is happening, but wonder if it's something to do with the bounding region prior to routing. It doesn't seem right than an earlier sub-part of a route should change if destination alters either.

Alternate 52.638454, 1.2218689 and 52.638454, 1.2218689 as destination to see earlier section differ.

**Comment by vshcherb** [comment]
Indeed via 52.631462, 1.2312781: routing_time = "2158.2036" and default time is routing_time = "2178.0046"

Though the difference is small.

**Comment by vshcherb** [comment]
Though it is related https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing.xml#L968 to heuristicCoefficient 1.2. So we accept 20% slower routes in order to calculate routes faster. With HR=1.0 it calculates correct route.

**Comment by PinkDuck** [comment]
That's interesting. Is the time saved in working out the route less than that I expend going the longer way around? Once I'm on my journey, there's plenty of time to go for a fully checked outcome through the re-routing capability. I doubt the initial outbound start would be affected much by that.

**Comment by sonora** [comment]
Well, the initial segment may well matter (we cannot generally simply "guess" it). That would lead to the funny situation where you are started off down one way, and then are maybe asked to turn right around retracing to where you were already... it would look like we "keep you busy" while we calculate ... surely not the desired solution either ... :)

Overall it is a valid question if you save more time by having a faster calculation at the risk of maybe not following the absolute fastest route (which in most cases you would still get with the current, faster mechanism).  Hence that can only be answered statistically, and your device speed is a critical factor in the answer. Not sure if it is worth to intelligently tweak the H value based on something like maybe straight line distance between starting point and destination or something. It would certainly not contribute to making things more transparent or debuggable here... and may only help rather little in some cases, while slowing down calculation for many other cases withou benefit. A common quagmire for any algorithm, by the way.

**Comment by PinkDuck** [comment]
I thought a bit more about this heuristic element overnight and just feel now that end users simply don't want non-optimal routes (it's why they want the device to do it), or if they are willing to accept that then perhaps for smaller excess distance than 20% in the case of pedestrian routing. There's not much you can do about slow performing devices, though perhaps disable heuristic if the overall journey length does not look particularly lengthy? In my example case, that's just a 5 km walk. Except the excess distance becomes more an issue for longer walks where the impact is greater.

The thing that bothered me here more was how the route itself changed with slight destination variation when the sub-section before has the same start/via points. When walking the route myself I chose to ignore the direction with the sense that I was right and the app was wrong, which afterwards turned out to be the case. That sort of mistrust in the app from user experience is not good.

**Comment by sonora** [comment]
Yeah, good point. But Victor would have to comment if anything could be done here. From my personal experience over the years it looks like occurrences where it matters seem rather rare, though.

**Comment by vshcherb** [comment]
We completely removed HR > 1 for car profile, though for bicycle profile it is very important to have it in many many cases it produces good result and most especially HR=1 doesn't work for routes around 100 km. It just very slow and crashes the application.

**Comment by xmd5a2** [comment]
With transit point:
```
Route <route>: routing_time=2323,2, vehicle=pedestrian, base=false, start_lat=52,62802, start_lon=1,23475, target_lat=52,64365, target_lon=1,21101, complete_distance=2987,0, complete_time=2763,3, native=true
Route <metrics>: tilesPerSec=6623,7, segmentsPerSec=156428,3
```
Without:
```
Route <route>: routing_time=2324,2, vehicle=pedestrian, base=false, start_lat=52,62802, start_lon=1,23475, target_lat=52,64365, target_lon=1,21101, complete_distance=2942,4, complete_time=2723,2, native=true
Route <metrics>: tilesPerSec=6987,8, segmentsPerSec=136970,1
```
